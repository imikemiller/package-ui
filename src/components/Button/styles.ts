import { css, SerializedStyles } from "@emotion/react";
import { Theme } from "../../theme";

/**
 * Function that accepts a theme and returns an array of style objects or maps of style objects
 *
 * Eg Style Object:
 *
 * css`
    color:"red"
  `;
 *
 * Eg Map of Style Objects:
 *
 * const buttonMode = {
    primary: css`
      color: "red";
    `,
    secondary: css`
      color: "blue"
    `,
  }
 *
 */
const styles = (
  theme: Theme,
): ({ [key: string]: SerializedStyles } | SerializedStyles)[] => {
  /**
     * Common button styles
     */
  const buttonBase = css`
      font-family: ${theme.fonts.button};
      font-weight: ${theme.fontWeights[2]};
      border: 0;
      border-radius: ${theme.borders.radius};
      cursor: pointer;
      display: inline-block;
      line-height: ${theme.lineHeights.button};
    `;

  /**
     * Primary/Secondary button variants
     */
  const buttonMode = {
    primary: css`
        ${buttonBase}
        color: ${theme.colors.primaryText};
        background-color: ${theme.colors.primary};
      `,
    secondary: css`
        ${buttonBase}
        color: #${theme.colors.secondaryText};
        background-color: transparent;
        border:${theme.borders.light} solid ${theme.colors.muted};
        box-shadow: ${theme.boxShadow.base};
      `,
  };

  /**
     * Size variants
     */
  const buttonSize = {
    small: css`
        font-size: ${theme.fontSizes[0]};
        padding: 10px 16px;
      `,
    medium: css`
        font-size: ${theme.fontSizes[1]};
        padding: 11px 20px;
      `,
    large: css`
        font-size: ${theme.fontSizes[2]};
        padding: 12px 24px;
      `,
  };

  return [buttonSize, buttonMode];
};

export default styles;
