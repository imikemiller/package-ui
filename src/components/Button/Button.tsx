/** @jsxRuntime classic */
/** @jsx jsx */
import { css, jsx } from "@emotion/react";
import React from "react";
import { useTheme } from "../../theme";
import styles from "./styles";

export interface ButtonProps {
  /**
   * Is this the principal call to action on the page?
   */
  mode?: "primary" | "secondary";
  /**
   * How large should the button be?
   */
  size?: "small" | "medium" | "large";
  /**
   * Button contents
   */
  label: string;
  /**
   * Optional click handler
   */
  onClick?: () => void;
}

/**
 * Primary UI component for user interaction
 */
export const Button: React.FC<ButtonProps> = ({
  mode = "primary",
  size = "medium",
  label,
  ...props
}) => {
  const theme = useTheme();
  const [buttonSize, buttonMode] = styles(theme);
  return (
    <button
      css={css`
        ${buttonMode[mode]}
        ${buttonSize[size]}
      `}
      {...props}
    >
      {label}
    </button>
  );
};
