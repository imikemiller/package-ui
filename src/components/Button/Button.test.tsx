import { snapshotTestFactory } from "../../utils/snapshotTestFactory";
import { Button, ButtonProps } from "./Button";

const props: ButtonProps[] = [
  {
    mode: "secondary",
    size: "small",
    label: "test",
  },
  {
    size: "medium",
    label: "test",
  },
  {
    size: "large",
    label: "test",
  },
];
snapshotTestFactory(Button, props);
