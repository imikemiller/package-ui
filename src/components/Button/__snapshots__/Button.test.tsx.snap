// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should match snapshot 1`] = `
<div>
  <button
    class="css-1t2kyz2-Button"
  >
    test
  </button>
</div>
`;

exports[`Button should match snapshot 2`] = `
<div>
  <button
    class="css-29z8yw-Button"
  >
    test
  </button>
</div>
`;

exports[`Button should match snapshot 3`] = `
<div>
  <button
    class="css-1gzd5ww-Button"
  >
    test
  </button>
</div>
`;
